### ISSUE TITLE
Snort blocking controlplane connections

### ISSUE THREAD

**Issue by zeroZshadow** [tailscale/tailscale#4582]
### What is the issue?

Somewhere recently the control plane has started reponding differenly, causing my installation of SNORT to block the traffic from it.
The blocking reason always seems to be `NO CONTENT-LENGTH OR TRANSFER-ENCODING IN HTTP RESPONSE`.

### Steps to reproduce

* Run [snort](https://www.snort.org/) on your router with the latest default block lists
* Execute `tailscale netcheck`
* Notice that the dial the ip for the controlplane has an i/o timeout.

### OS

Linux, Windows, Android

### OS version

The reproduction test was done on latest Raspbian

### Tailscale version

1.24.2

**Comment by DentonGentry** [comment]
1.24.x implements a new transport for the control protocol, switching to NOISE from Naclbox. It attempts to make a connection to port 80 first, and will only try port 443 if port 80 fails to exchange data. Earlier client versions never attempted to use port 80.

**Comment by danderson** [comment]
It looks to me like snort doesn't understand `HTTP 101 Switching Protocols` responses, which is what we do to signal that the HTTP request. Such responses implicitly have no response body, since the whole point is to switch protocols away from HTTP.

We could add more useless headers to the response, but I'm wondering what the standards say to expect...

**Comment by danderson** [comment]
cc @bradfitz aside from "we should obviously do whatever makes real implementation happy", can you opine? Are we actually messing up the HTTP 101 response by not specifying a content length/encoding, or should we complain to snort that their heuristic is wrong? :)

Either way, for actual use, this _should_ not be an issue unless your snort installation is also doing HTTPS MITM, because we built a fallback into this protocol where it'll retry the dial over HTTPS if HTTP seems to be failing for proxy/IDS reasons. We should definitely still fix it, but @zeroZshadow can you confirm that tailscale 1.24.2 still works correctly overall, despite the snort failures? If not, we may also have a bug in our fallback logic.

**Comment by zeroZshadow** [comment]
Looking at the definition, Content-Length and Transfer-Encoding should never be on 1xx responses, so this feels like a Snort mistake.

@danderson 
With snort blocking the traffic, tailscale is not functioning. The windows tray app mentions the windows service isnt running and `tailscale status` will print a NoState.

**Comment by danderson** [comment]
Hmm. If snort terminates the port 80 connection, we should still succeed with a control connection over TLS on :443, so I'm wondering why that's not happening... Do your snort logs say anything about connections to https://controlplane.tailscale.com ? We had some bugs with the TLS fallback in <1.14.2, but in .2 it should all be fixed...

**Comment by zeroZshadow** [comment]
I think the problem is that snort doesn't just terminal port 80, it blocks all traffic from the originating ip for an hour.

**Comment by danderson** [comment]
aah, okay. So we don't even get a chance to try the fallback. That makes sense.

I'm worried that if we try to add Content-Length=0 to our 101 response, it'll make other IDSes angry because we're violating the spec. It looks like snort rules are just in plaintext, so I guess I could try to figure out how a patch for snort to make it handle 101 responses properly. But I don't know snort at all, so it'll probably take a while.

**Comment by danderson** [comment]
For context on why, btw: the new control protocol uses the Noise protocol framework, which has nicer security properties for us than TLS. To avoid double-encryption, we try to upgrade HTTP->Noise over port 80 first, and then if that fails we fall back to Noise over TLS for maximum compatibility. It didn't occur to me that even trying the port 80 option would make mitm boxes angry :(

**Comment by zeroZshadow** [comment]
Yeah what tailscale is doing it very sensible, and seems to be correct as well. I'm curious if Snort responds similarly to websocket protocol changes on port 80.

I've disable the snort rule on my end for now, It's a home setup after in my case, so this work around is acceptable. I think Snort allows for reporting for false positives, but I've never done so in the past. (I've simply installed the package on pfSense and never touched it again).

**Comment by danderson** [comment]
If you have a way to report false positives to snort, it'd be great to tell them that their rule should allow a missing content length/encoding on 101 responses, since that's the standards-compliant thing.

We were just wondering if websockets stuff is also just broken on your network, because yeah WS does exactly the same protocol switch. Maybe that's just being saved because ~everything is HTTPS now, so snort's buggy rule can't see that traffic?...

**Comment by danderson** [comment]
To confirm, so we can try to set this up in a lab and poke at snort: which version of snort are you running? (I'm assuming its rule engine etc. changes over time, so I want a bug-for-bug match)

**Comment by zeroZshadow** [comment]
I'm using Snort 4.1.5_3 on pfSense
I've just tried to have the alert trigger on http://websocketstest.com/ but I couldn't get it to reproduce.
I'm currently trying to get tailscale to trigger this again and get a packet capture recording. No promisses on that since I've never done this before.

**Comment by zeroZshadow** [comment]
Correction, it seems it took a bit for the rules list to update. The alert ALSO triggers on http://websocketstest.com/
So any http protocol upgrade is going to be blocked it seems. I'll see if I can report this false positive.

**Comment by zeroZshadow** [comment]
I've reported the false positive and will report back if there is any update

**Comment by danderson** [comment]
Thanks! Hopefully we can just get Snort to fix the rules, then we don't have to try and figure out workarounds :)

**Comment by citadella** [comment]
FYI: I was having this issue and found a workaround (thus far) was to add FW Aliases for tailscale.com, login.tailscale.com, & controlplane.tailscale.com, clear Snort blocklists, and that worked. Not an ideal solution, but pretty a pretty low-effort workaround...

**Comment by DentonGentry** [comment]
I cannot find an open Snort issue about `101 switching protocols`
